### DISCUSSION TITLE
Cumulative Distance Property

### DISCUSSION THREAD

**Discussion by judeurban** [JSBSim-Team/jsbsim#831] [Q&A]
Hello, all.

I'm searching for a **"cumulative distance since origin"** parameter. Please note that this is different than the *norm* since the origin. It is the distance accumulated after each discrete timestep from one point to the next.

![distance](https://user-images.githubusercontent.com/73368004/217102453-f89598f0-90fc-4a3d-a5c3-5b96ecea2794.png)

I've found the parameters below using [JSB Reference Manual](https://jsbsim.sourceforge.net/JSBSimReferenceManual.pdf), but they provide the norm since $t = 0$, not the cumulative distance.

- `position/distance-from-start-lat-mt`
- `position/distance-from-start-lon-mt`
- `position/distance-from-start-mag-mt`

I've also attempted to utilize the [FGLocation's](https://jsbsim-team.github.io/jsbsim/classJSBSim_1_1FGLocation.html) [`GetDistanceTo()`](https://jsbsim-team.github.io/jsbsim/classJSBSim_1_1FGLocation.html#afa50b3c5c4e01a2b203a9c30dcf67414) method to compute each distance myself. In theory, this is possible because I keep track of the latitude/longitude from the previous time step.

```c++
// get the current position from the FDM
currentLatitude = FDM.GetPropertyValue("position/lat-gc-deg");
currentLongitude = FDM.GetPropertyValue("position/long-gc-deg");

// provide FGLocation object with this information
FGLocation.SetLatitude(currentLatitude * DEG2RAD);
GLocation.SetLongitude(currentLongitude * DEG2RAD);

// use the GetDistanceTo() method to compue the distance. This only returns NaN.
distanceFromOrigin += FGLocation.GetDistanceTo(position.previousLongitude * DEG2RAD, position.previousLongitude * DEG2RAD);

// overwrite the previous position
previousLatitude = currentLatitude;
previousLongitude = currentLongitude;
```

Is there a property that does what I am looking for, or should I compute this information manually? Depending on your answer, perhaps this thread will be miscategorized in the Q&A section and would be better suited in the feature request section.

Thank you!

**Comment by seanmcleod** [comment]
I'm not aware of a property off-hand which tracks the cumulative distance. But if your code is running after each timestep then one option is to calculate the distance moved during each timestep and accumulate that yourself using ECEF coordinates. 

```c++
  PropertyManager->Tie("position/ecef-x-ft", this, eX, (PMF)&FGPropagate::GetLocation);
  PropertyManager->Tie("position/ecef-y-ft", this, eY, (PMF)&FGPropagate::GetLocation);
  PropertyManager->Tie("position/ecef-z-ft", this, eZ, (PMF)&FGPropagate::GetLocation);
```

**Comment by seanmcleod** [comment]
@judeurban what approach did you go with?

**Reply by judeurban** [reply]
Since I'm using a combination of python and C++ for this project, I decided to just accumulate the range in Python. See the function below. This was called after each iteration to compute the distance. For anyone reading this in the future, this could of course be duplicated quite easily in C++. The function uses the [Haversine Formula](https://en.wikipedia.org/wiki/Haversine_formula). Here is [another implementation](https://stackoverflow.com/questions/4913349/haversine-formula-in-python-bearing-and-distance-between-two-gps-points) on stack overflow.

```python

EARTH_RADIUS_FEET = 20_925_721

def distance_between_two_points(p0: tuple, p1: tuple) -> float:
    """Compute a distance between two lat/lon points in feet

    Args:
        p0 (tuple): (Latitude, Longitude) for point 0
        p1 (tuple): (Latitude, Longitude) for point 1

    Returns:
        float: Distance between these two points
    """

    lat1 = radians(p0[0])
    lon1 = radians(p0[1])
    lat2 = radians(p1[0])
    lon2 = radians(p1[1])

    delta_lat = lat2 - lat1
    delta_lon = lon2 - lon1

    a = sin(delta_lat / 2)**2 + cos(lat1) * cos(lat2) * sin(delta_lon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = EARTH_RADIUS_FEET * c

    return distance
```

**Comment by seanmcleod** [comment]
@judeurban any reason you didn't use the ECEF coordinates? Much simpler Pythagoras formula to calculate distance compared to the archaversine (inverse haversine) formula you've used.

The haversine is calculating the great circle distance which isn't necessarily the path taken by the vehicle, although realistically given you're doing this each timestep you probably wouldn't notice a difference since the delta distances are so small. It also assumes a spherical earth based on a constant earth radius as opposed to a WGS84 ellipsoid model of the earth. Both these issues are avoided by using ECEF coordinates.
